### PR TITLE
[client/network] Add support  for `/wss` addresses

### LINKED ISSUE #12024: /wss/p2p protocol not supported?
I would like to verify if a node is configure properly for usages of substrate-connect. I usually starts with a Substrate node to verify it can indeed connect to the new node. However it seems like it just doesn't work with `/wss/p2p` multi addr at all?

I tried this
`docker run --rm parity/polkadot:v0.9.27 --reserved-only --reserved-nodes=/dns/polkadot-connect-7.parity.io/tcp/443/wss/p2p/12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o -lsub-libp2p=trace`

and getting error

```
2022-08-13 08:20:50.703 TRACE tokio-runtime-worker sub-libp2p: Libp2p => Reach failure for Some(PeerId("12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o")) through "/dns/polkadot-connect-7.parity.io/tcp/443/wss/p2p/12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o": Other(Custom { kind: Other, error: Other(A(A(A(A(MultiaddrNotSupported("/ip4/34.141.46.208/tcp/443/wss/p2p/12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o")))))) })    
2022-08-13 08:20:50.703 TRACE tokio-runtime-worker sub-libp2p: Libp2p => Reach failure for Some(PeerId("12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o")) through "/dns/polkadot-connect-0.parity.io/tcp/443/wss/p2p/12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o": Other(Custom { kind: Other, error: Other(A(A(A(A(MultiaddrNotSupported("/ip4/35.234.84.149/tcp/443/wss/p2p/12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o")))))) })    
2022-08-13 08:20:50.703 TRACE tokio-runtime-worker sub-libp2p: Libp2p => Dial failure for PeerId("12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o")    
2022-08-13 08:20:50.703 TRACE tokio-runtime-worker sub-libp2p: PSM <= Dropped(12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o, SetId(0))    
2022-08-13 08:20:50.703 TRACE tokio-runtime-worker sub-libp2p: Libp2p => Failed to reach PeerId("12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o"): An error occurred while negotiating the transport protocol(s) on a connection: [("/dns/polkadot-connect-7.parity.io/tcp/443/wss/p2p/12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o", Other(Custom { kind: Other, error: Other(A(A(A(A(MultiaddrNotSupported("/ip4/34.141.46.208/tcp/443/wss/p2p/12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o")))))) })), ("/dns/polkadot-connect-0.parity.io/tcp/443/wss/p2p/12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o", Other(Custom { kind: Other, error: Other(A(A(A(A(MultiaddrNotSupported("/ip4/35.234.84.149/tcp/443/wss/p2p/12D3KooWEPmjoRpDSUuiTjvyNDd8fejZ9eNWH5bE965nyBMDrB4o")))))) }))].   
```

### PR DESCRIPTION
**Pull request by melekes**
Because WSS transport requires addresses w/ DNS names, I had to construct it and use alongside DNS transport (which contains another instance of TCP transport). WSS transport itself also needs an inner instance of DNS transport to resolve DNS names.

Refs https://github.com/libp2p/rust-libp2p/issues/3330
Closes #12024

### CODE REVIEW

File: client/network/src/transport.rs [hunk 1]
```diff
@@ -55,17 +55,25 @@ pub fn build_transport(
...
 		EitherTransport::Left(if let Ok(dns) = dns_init {
-			EitherTransport::Left(dns)
+			let tcp_trans = tcp::tokio::Transport::new(tcp_config.clone());
+			let desktop_trans = websocket::WsConfig::new(tcp_trans)
+				.or_transport(tcp::tokio::Transport::new(tcp_config.clone()));
+			let dns_init = dns::TokioDnsConfig::system(desktop_trans);
+			// Secure Websocket transport needs unresolved addresses, so we join DNS transport with
+			// yet another instance of Websocket transport.
+			EitherTransport::Left(
+				websocket::WsConfig::new(dns).or_transport(dns_init.expect("same config to work")),
```
Comment (dmitry-markin) [inline client/network/src/transport.rs:70]: `dns` here already contains an instance of `WsConfig`. Intuitively, `DnsConfig` required by `WsConfig::new` should only provide DNS and TCP transport. May be it makes sense separating DNS carrying TCP+WS and "bare" DNS for WSS?

Comment (dmitry-markin) [reply client/network/src/transport.rs:70]: Also, I don't know internals of libp2p good enough to reason with confidence, but may be `WsConfig` based on `DnsConfig` is capable of handling raw `ip4`/`ip6` addresses and we don't need two instances of WS transport?

Comment (melekes) [reply client/network/src/transport.rs]: > May be it makes sense separating DNS carrying TCP+WS and "bare" DNS for WSS?

good point. done in https://github.com/paritytech/substrate/pull/13152/commits/756ef866b2fc8b78c93568078310af4b7ce9907c

**Comment by melekes** [comment]
bot rebase

File: client/network/src/transport.rs (hunk 1 above)
Comment (dmitry-markin) [reply client/network/src/transport.rs]: > Also, I don't know internals of libp2p good enough to reason with confidence, but may be `WsConfig` based on `DnsConfig` is capable of handling raw `ip4`/`ip6` addresses and we don't need two instances of WS transport?

Did you try running without second instance of WS transport for bare IP addresses?

**Review by bkchr — APPROVED** [review]

File: client/network/src/transport.rs (hunk 1 above)
Comment (melekes) [reply client/network/src/transport.rs]: > Did you try running without second instance of WS transport for bare IP addresses?

Trying right now. It should work in theory

Comment (melekes) [reply client/network/src/transport.rs]: It works 👍

**Comment by melekes** [comment]
bot merge

**Pull request merged**